File: LeetCode/LeetCode Practice Problems/1525- Number of Good Ways to Split a String.cpp
class Solution {
public:
    int numSplits(string s) {
        int diff = 0 , ans = 0;
        int n = s.length();
        vector<int> f1(26,0),f2(26,0);
        vector<int> l(n) , r(n);
        int i = 0;
        
        for(char c: s) {
            f1[c - 'a']++;
            if(f1[c - 'a'] == 1) diff++;
            
            l[i++] = diff;
        }
        diff = 0;
        
        for(int i=n-1;i>=0;i--) {
            f2[s[i] - 'a']++;
            if(f2[s[i] - 'a'] == 1) diff++;
            
            r[i] = diff;
        }
        
        for(int i=0;i<n-1;i++) {
            if(l[i] == r[i+1]) ans++;
        }
        
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/1525-Number of Good Ways to Split a String.java
class Solution {
    public int numSplits(String s) {
        int freq[] = new int[26];
        int distinct = 0;
        
        for(char c: s.toCharArray()) {
            freq[c - 'a']++;
            if(freq[c - 'a'] == 1) 
                distinct++;
        }
        
        int leftDistinct = 0 , rightDistinct = distinct;
        int ans = 0 , idx = 0;
        int freq2[] = new int[26];
        
        while(leftDistinct <= rightDistinct && idx < s.length()) {
            char c = s.charAt(idx);
            
            freq[c - 'a']--;
            if(freq[c - 'a'] == 0) {
                rightDistinct--;
            }
            
            freq2[c - 'a']++;
            if(freq2[c - 'a'] == 1) {
                leftDistinct++;
            }
            
            ans += (leftDistinct == rightDistinct)?1:0;
            idx++;
        }
        
        return ans;
    }
}
